### PR TITLE
Infer HoC props

### PR DESCRIPTION
**Pull request by AndrewMusgrave**
### WHY are these changes introduced?

Improved HoC usage
* withSticky
* withAppProvider 
* withContext
* withRef

##### Old
```tsx
export default withAppProvider<Props>()(Component)
```

##### New
```tsx
export default withAppProvider()(Component)
```

##### Old

```tsx
export default compose<Props>(
  withContext<Props, DropZoneContext>(Consumer),
  withAppProvider<Props>(),
  withRef<Props>(),
)(FileUpload);
```

##### New

```tsx
export default compose<Props>(
  withContext<DropZoneContext>(Consumer),
  withAppProvider(),
  withRef(),
)(FileUpload);
```

## 

### How to 🎩

This is a pretty big change so we'll want to test the components making use of different HoC.
Here's small list of some components you'll want to test making use of some HoC's

#### withSticky

* Scrollable

#### withAppProvider 

* Button
* DatePicker
* OptionList
* most components 😄 

#### withContext

* DropZone.FileUpload
* ResourceList.FilterControl
* ResourceList.Item

#### withRef

* DropZone.FileUpload
* UnstyledLink

### CODE REVIEW

**Comment by BPScott** [comment]
I'm not going to pretend to understand the TS intricacies that let this work but I love the sight of all those

```diff
- export default withAppProvider<Props>()(Thing);
+ export default withAppProvider()(Thing);
```

changes!

File: src/components/AppProvider/utilities/withAppProvider/withAppProvider.tsx [hunk 1]
```diff
@@ -23,13 +23,13 @@ export interface WithAppProviderProps {
   };
 }
 
-export default function withAppProvider<OwnProps>() {
-  return function addProvider<C>(
+export default function withAppProvider() {
+  return function addProvider<OriginalProps, C>(
```
Comment (dleroux) [inline src/components/AppProvider/utilities/withAppProvider/withAppProvider.tsx]: C is Context in this case correct?

Comment (AndrewMusgrave) [reply src/components/AppProvider/utilities/withAppProvider/withAppProvider.tsx]: Yup 👍

**Comment by AndrewMusgrave** [comment]
> This is great. What version of Typescript made this possible?

Not sure

> I couldn't build-consumer because of icons to test on web. I can't see see the resulting build impacting web but going with lessons learned.

I think it'll be best to wait as well, until we can 🎩 in web

**Comment by AndrewMusgrave** [comment]
### Steps to 🎩 in web

#### You'll want to be on master in `web` and hoc-infer-types in `polaris-react`
```bash
# 1
dev cd web
# 2
dev down && dev up && yarn add -D @Shopify/polaris-icons
# 3
dev cd polaris-react
# 4
dev up && yarn build-consumer web
# 5
dev cd web && dev server
```

**Pull request merged**